i3py/drivers/yokogawa/gs200.py

```diff
@@ -0,0 +1,231 @@
+
+VOLTAGE_RESOLUTION = {10e-3: 1e-7,
+                      100e-3: 1e-6,
+                      1.0: 1e-5,
+                      10: 1e-4,
+                      30: 1e-3}
+
+CURRENT_RESOLUTION = {1e-3: 1e-8,
+                      10e-3: 1e-7,
+                      100e-3: 1e-6,
+                      200e-3: 1e-6}
+
+
+class GS200(DCPowerSource, IEEEInternalOperations,
+            IEEEStatusReporting, IEEEOperationComplete,
+            IEEEOptionsIdentification, IEEEStoredSettings,
+            SCPIErrorReading):
+    """Driver for the Yokogawa GS200 DC power source.
+
+    Because the over-voltage (current) protection is always enabled in
+    current (voltage) mode, they basically act as limits and fully the same
+    role as target voltage (current) for a power source lacking mode selection.
+    As a consequence they are implemented in the same way.
+
+    Notes
+    -----
+    - the measurement option is not yet supported.
+    - add support for programs
+    - add RS232 support
+
+    """
+    __version__ = '0.1.0'
+
+    PROTOCOLS = {'GPIB': [{'resource_class': 'INSTR'}],
+                 'USB': [{'resource_class': 'INSTR',
+                          'manufacturer_id': '0xB21',
+                          'model_code': '0x39'}],
+                 'TCPIP': [{'resource_class': 'INSTR'}]
+                 }
+
+    DEFAULTS = {'COMMON': {'read_termination': '\n',
+                           'write_termination': '\n'}}
+
+    outputs = channel((0,))
+
+    with outputs as o:
+        #: Preferential working mode for the source. In voltage mode, the
+        #: source tries to work as a voltage source, the current settings is
+        #: simply used to protect the sample. In current mode it is the
+        #: opposite. Changing the mode cause the output to be disabled.
+        o.mode = Str(getter=':SOUR:FUNC?',
+                     setter=':SOUR:FUNC {}',
+                     mapping={'voltage': 'VOLT', 'current': 'CURR'},
+                     discard={'features': ('enabled',
+                                           'voltage', 'voltage_range',
+                                           'current', 'current_range'),
+                              'limits': ('voltage', 'current')})
+
+        o.enabled = set_feat(getter=':OUTP?', setter=':OUTP {}',
+                             mapping={False: '0', True: '1'})
+
+        o.voltage = set_feat(
+            getter=conditional('":SOUR:LEV?" if driver.mode == "voltage" '
+                               'else ":SOUR:PROT:VOLT?"', default=True),
+            setter=conditional('":SOUR:LEV {}" if driver.mode == "voltage" '
+                               'else ":SOUR:PROT:VOLT {}"', default=True),
+            limits='voltage')
+
+        o.voltage_range = set_feat(getter=True,
+                                   setter=':SOUR:RANG {}',
+                                   checks=(None, 'driver.mode == "voltage"'),
+                                   values=(10e-3, 100e-3, 1.0, 10.0, 30.0),
```

Comment:
You could use your VOLTAGE_RESOLUTION keys instead of rewriting it (idem below)

Reply:
Will do